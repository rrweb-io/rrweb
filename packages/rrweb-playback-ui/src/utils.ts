declare global {
  interface Document {
    mozExitFullscreen: Document['exitFullscreen'];
    webkitExitFullscreen: Document['exitFullscreen'];
    msExitFullscreen: Document['exitFullscreen'];
    webkitIsFullScreen: Document['fullscreen'];
    mozFullScreen: Document['fullscreen'];
    msFullscreenElement: Document['fullscreen'];
  }

  interface HTMLElement {
    mozRequestFullScreen: Element['requestFullscreen'];
    webkitRequestFullscreen: Element['requestFullscreen'];
    msRequestFullscreen: Element['requestFullscreen'];
  }
}

export function inlineCss(cssObj: Record<string, string>): string {
  let style = '';
  Object.keys(cssObj).forEach((key) => {
    style += `${key}: ${cssObj[key]};`;
  });
  return style;
}

function padZero(num: number, len = 2): string {
  let str = String(num);
  const threshold = Math.pow(10, len - 1);
  if (num < threshold) {
    while (String(threshold).length > str.length) {
      str = '0' + num;
    }
  }
  return str;
}

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
export function formatTime(ms: number): string {
  if (ms <= 0) {
    return '00:00';
  }
  const hour = Math.floor(ms / HOUR);
  ms = ms % HOUR;
  const minute = Math.floor(ms / MINUTE);
  ms = ms % MINUTE;
  const second = Math.floor(ms / SECOND);
  if (hour) {
    return `${padZero(hour)}:${padZero(minute)}:${padZero(second)}`;
  }
  return `${padZero(minute)}:${padZero(second)}`;
}

export function openFullscreen(el: HTMLElement): Promise<void> {
  if (el.requestFullscreen) {
    return el.requestFullscreen();
  } else if (el.mozRequestFullScreen) {
    /* Firefox */
    return el.mozRequestFullScreen();
  } else if (el.webkitRequestFullscreen) {
    /* Chrome, Safari and Opera */
    return el.webkitRequestFullscreen();
  } else if (el.msRequestFullscreen) {
    /* IE/Edge */
    return el.msRequestFullscreen();
  }
}

export function exitFullscreen(): Promise<void> {
  if (document.exitFullscreen) {
    return document.exitFullscreen();
  } else if (document.mozExitFullscreen) {
    /* Firefox */
    return document.mozExitFullscreen();
  } else if (document.webkitExitFullscreen) {
    /* Chrome, Safari and Opera */
    return document.webkitExitFullscreen();
  } else if (document.msExitFullscreen) {
    /* IE/Edge */
    return document.msExitFullscreen();
  }
}

export function isFullscreen(): boolean {
  return (
    document.fullscreen ||
    document.webkitIsFullScreen ||
    document.mozFullScreen ||
    document.msFullscreenElement
  );
}

export function onFullscreenChange(handler: () => unknown): () => void {
  document.addEventListener('fullscreenchange', handler);
  document.addEventListener('webkitfullscreenchange', handler);
  document.addEventListener('mozfullscreenchange', handler);
  document.addEventListener('MSFullscreenChange', handler);

  return () => {
    document.removeEventListener('fullscreenchange', handler);
    document.removeEventListener('webkitfullscreenchange', handler);
    document.removeEventListener('mozfullscreenchange', handler);
    document.removeEventListener('MSFullscreenChange', handler);
  };
}

export function typeOf(
  obj: unknown,
):
  | 'boolean'
  | 'number'
  | 'string'
  | 'function'
  | 'array'
  | 'date'
  | 'regExp'
  | 'undefined'
  | 'null'
  | 'object' {
  const toString = Object.prototype.toString;
  const map = {
    '[object Boolean]': 'boolean',
    '[object Number]': 'number',
    '[object String]': 'string',
    '[object Function]': 'function',
    '[object Array]': 'array',
    '[object Date]': 'date',
    '[object RegExp]': 'regExp',
    '[object Undefined]': 'undefined',
    '[object Null]': 'null',
    '[object Object]': 'object',
  };
  return map[toString.call(obj)];
}
