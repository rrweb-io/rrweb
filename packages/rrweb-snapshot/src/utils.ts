import { INode, MaskInputFn, MaskInputOptions } from './types';

export function isElement(n: Node | INode): n is Element {
  return n.nodeType === n.ELEMENT_NODE;
}

export function isShadowRoot(n: Node): n is ShadowRoot {
  const host: Element | null = (n as ShadowRoot)?.host;
  return Boolean(host && host.shadowRoot && host.shadowRoot === n);
}

export function maskInputValue({
  maskInputOptions,
  tagName,
  type,
  value,
  maskInputFn,
}: {
  maskInputOptions: MaskInputOptions;
  tagName: string;
  type: string | number | boolean | null;
  value: string | null;
  maskInputFn?: MaskInputFn;
}): string {
  let text = value || '';
  if (
    maskInputOptions[tagName.toLowerCase() as keyof MaskInputOptions] ||
    maskInputOptions[type as keyof MaskInputOptions]
  ) {
    if (maskInputFn) {
      text = maskInputFn(text);
    } else {
      text = '*'.repeat(text.length);
    }
  }
  return text;
}

const ORIGINAL_ATTRIBUTE_NAME = '__rrweb_original__';
type PatchedGetImageData = {
  [ORIGINAL_ATTRIBUTE_NAME]: CanvasImageData['getImageData'];
} & CanvasImageData['getImageData'];

export function is2DCanvasBlank(canvas: HTMLCanvasElement): boolean {
  const ctx = canvas.getContext('2d');
  if (!ctx) return true;

  const chunkSize = 50;

  // get chunks of the canvas and check if it is blank
  for (let x = 0; x < canvas.width; x += chunkSize) {
    for (let y = 0; y < canvas.height; y += chunkSize) {
      const getImageData = ctx.getImageData as PatchedGetImageData;
      const originalGetImageData =
        ORIGINAL_ATTRIBUTE_NAME in getImageData
          ? getImageData[ORIGINAL_ATTRIBUTE_NAME]
          : getImageData;
      // by getting the canvas in chunks we avoid an expensive
      // `getImageData` call that retrieves everything
      // even if we can already tell from the first chunk(s) that
      // the canvas isn't blank
      const pixelBuffer = new Uint32Array(
        originalGetImageData.call(
          ctx,
          x,
          y,
          Math.min(chunkSize, canvas.width - x),
          Math.min(chunkSize, canvas.height - y),
        ).data.buffer,
      );
      if (pixelBuffer.some((pixel) => pixel !== 0)) return false;
    }
  }
  return true;
}
