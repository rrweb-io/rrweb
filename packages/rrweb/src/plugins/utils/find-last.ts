export function findLast<T>(
  array: Array<T>,
  predicate: (value: T) => boolean,
): T | undefined {
  const length = array.length;
  for (let i = length - 1; i >= 0; i -= 1) {
    if (predicate(array[i])) {
      return array[i];
    }
  }
}
