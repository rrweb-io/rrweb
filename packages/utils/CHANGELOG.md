# @rrweb/utils

## 2.0.0-alpha.18

## 2.0.0-alpha.17

### Patch Changes

- [#1509](https://github.com/rrweb-io/rrweb/pull/1509) [`be6bf52`](https://github.com/rrweb-io/rrweb/commit/be6bf52c248c35de1b3491e3a3440ff61f876414) Thanks [@Juice10](https://github.com/Juice10)! - Reverse monkey patch built in methods to support LWC (and other frameworks like angular which monkey patch built in methods).
