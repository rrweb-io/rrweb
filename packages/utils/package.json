{
  "name": "@rrweb/utils",
  "version": "2.0.0-alpha.17",
  "publishConfig": {
    "access": "public"
  },
  "keywords": [
    "rrweb",
    "@rrweb/utils"
  ],
  "scripts": {
    "dev": "vite build --watch",
    "build": "tsc -noEmit && vite build",
    "check-types": "tsc -noEmit",
    "prepublish": "npm run build",
    "lint": "yarn eslint src/**/*.ts"
  },
  "homepage": "https://github.com/rrweb-io/rrweb/tree/main/packages/@rrweb/utils#readme",
  "bugs": {
    "url": "https://github.com/rrweb-io/rrweb/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/rrweb-io/rrweb.git"
  },
  "license": "MIT",
  "type": "module",
  "main": "./dist/utils.umd.cjs",
  "module": "./dist/utils.js",
  "unpkg": "./dist/utils.umd.cjs",
  "typings": "dist/index.d.ts",
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/utils.js"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/utils.umd.cjs"
      }
    }
  },
  "files": [
    "dist",
    "package.json"
  ],
  "devDependencies": {
    "vite": "^5.2.8",
    "vite-plugin-dts": "^3.8.1"
  },
  "dependencies": {}
}
