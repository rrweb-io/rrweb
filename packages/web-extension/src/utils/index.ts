export function isFirefox(): boolean {
  return (
    (typeof window !== 'undefined' &&
      window.navigator?.userAgent.toLowerCase().includes('firefox')) ||
    false
  );
}

export function isInCrossOriginIFrame(): boolean {
  if (window.parent !== window) {
    try {
      void window.parent.location.origin;
    } catch (error) {
      return true;
    }
  }
  return false;
}

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;

export function formatTime(ms: number): string {
  if (ms <= 0) {
    return '00:00';
  }
  const hour = Math.floor(ms / HOUR);
  ms = ms % HOUR;
  const minute = Math.floor(ms / MINUTE);
  ms = ms % MINUTE;
  const second = Math.floor(ms / SECOND);
  if (hour) {
    return `${padZero(hour)}:${padZero(minute)}:${padZero(second)}`;
  }
  return `${padZero(minute)}:${padZero(second)}`;
}

function padZero(num: number, len = 2): string {
  let str = String(num);
  const threshold = Math.pow(10, len - 1);
  if (num < threshold) {
    while (String(threshold).length > str.length) {
      str = `0${num}`;
    }
  }
  return str;
}
