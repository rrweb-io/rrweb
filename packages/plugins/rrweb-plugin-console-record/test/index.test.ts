import { describe, it, beforeAll, afterAll, expect, vi } from 'vitest';
import { stringifySnapshots } from '../../../rrweb/test/utils';
import { createServer, ViteDevServer } from 'vite';
import * as puppeteer from 'puppeteer';
import type { Browser, Page } from 'puppeteer';
import type { eventWithTime } from '@rrweb/types';

export async function launchPuppeteer(
  options?: Parameters<(typeof puppeteer)['launch']>[0],
) {
  return await puppeteer.launch({
    headless: process.env.PUPPETEER_HEADLESS ? true : false,
    defaultViewport: {
      width: 1920,
      height: 1080,
    },
    args: ['--no-sandbox'],
    ...options,
  });
}

export function assertSnapshot(snapshots: eventWithTime[]) {
  expect(snapshots).toBeDefined();
  expect(stringifySnapshots(snapshots)).toMatchSnapshot();
}

describe('rrweb-plugin-console-record', () => {
  // vi.setConfig({ testTimeout: 120_000 });
  let server: ViteDevServer;
  let serverUrl: string;
  let browser: Browser;
  let page: Page;

  beforeAll(async () => {
    server = await createServer({
      preview: { port: 3000 },
      mode: 'test',
      // hmr calls `console.debug('[vite] connected')` and messes up our snapshots
      // so we disable it
      server: { hmr: false },
    });
    await server.listen();
    serverUrl = server.resolvedUrls!.local[0];
    browser = await launchPuppeteer();
    page = await browser.newPage();
  });

  afterAll(async () => {
    await browser.close();
    await server.close();
  });

  it('should handle recursive console messages', async () => {
    await page.goto(`${serverUrl}test/html/log.html`);

    await page.evaluate(() => {
      // Some frameworks like Vue.js use proxies to implement reactivity.
      // This can cause infinite loops when logging objects.
      let recursiveTarget = { foo: 'bar', proxied: 'i-am', proxy: null };
      let count = 0;

      const handler = {
        get(target: any, prop: any, ...args: any[]) {
          if (prop === 'proxied') {
            if (count > 9) {
              return;
            }
            count++; // We don't want out test to get into an infinite loop...
            console.warn(
              'proxied was accessed so triggering a console.warn',
              target,
            );
          }
          return Reflect.get(target, prop, ...args);
        },
      };

      const proxy = new Proxy(recursiveTarget, handler);
      recursiveTarget.proxy = proxy;

      console.log('Proxied object:', proxy);
    });

    // await waitForRAF(page);

    const snapshots = (await page.evaluate(
      'window.snapshots',
    )) as eventWithTime[];
    // The snapshots should containe 1 console log, not multiple.
    await assertSnapshot(snapshots);
  });

  it('should record console messages', async () => {
    await page.goto(`${serverUrl}test/html/log.html`);

    await page.evaluate(() => {
      // truthy assert does not log
      console.assert(0 === 0, 'should not log assert');
      // falsy assert does log
      console.assert(false, 'should log assert');
      console.count('count');
      console.countReset('count');
      console.debug('debug');
      console.dir('dir');
      console.dirxml('dirxml');
      console.group();
      console.groupCollapsed();
      console.info('info');
      console.log('log');
      console.table('table');
      console.time();
      console.timeEnd();
      console.timeLog();
      console.trace('trace');
      console.warn('warn');
      console.clear();
      console.log(new TypeError('a message'));
      const iframe = document.createElement('iframe');
      document.body.appendChild(iframe);
    });

    await page.frames()[1].evaluate(() => {
      console.log('from iframe');
    });

    const snapshots = (await page.evaluate(
      'window.snapshots',
    )) as eventWithTime[];
    await assertSnapshot(snapshots);
  });
});
