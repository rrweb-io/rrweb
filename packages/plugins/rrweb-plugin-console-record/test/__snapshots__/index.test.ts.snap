// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rrweb-plugin-console-record > should handle recursive console messages 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080,
      \\"captureAssets\\": {
        \\"objectURLs\\": true,
        \\"origins\\": false
      }
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {
              \\"lang\\": \\"en\\"
            },
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 5
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"module\\",
                      \\"src\\": \\"http://localhost:3030/@vite/client\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 6
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n\\\\n    \\",
                    \\"id\\": 7
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"charset\\": \\"UTF-8\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 8
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 9
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"name\\": \\"viewport\\",
                      \\"content\\": \\"width=device-width, initial-scale=1.0\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 10
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 11
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"http-equiv\\": \\"X-UA-Compatible\\",
                      \\"content\\": \\"ie=edge\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 12
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 13
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"title\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"Log record\\",
                        \\"id\\": 15
                      }
                    ],
                    \\"id\\": 14
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 16
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"src\\": \\"http://localhost:3030/test/html/index.ts\\",
                      \\"type\\": \\"module\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 17
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\",
                    \\"id\\": 18
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 3,
                \\"textContent\\": \\"\\\\n  \\",
                \\"id\\": 19
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n\\\\n\\",
                    \\"id\\": 21
                  }
                ],
                \\"id\\": 20
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"log\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:21:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"Proxied object:\\\\\\"\\",
          \\"\\\\\\"[object Object]\\\\\\"\\"
        ]
      }
    }
  }
]"
`;

exports[`rrweb-plugin-console-record > should record console messages 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080,
      \\"captureAssets\\": {
        \\"objectURLs\\": true,
        \\"origins\\": false
      }
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {
              \\"lang\\": \\"en\\"
            },
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 5
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"module\\",
                      \\"src\\": \\"http://localhost:3030/@vite/client\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 6
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n\\\\n    \\",
                    \\"id\\": 7
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"charset\\": \\"UTF-8\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 8
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 9
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"name\\": \\"viewport\\",
                      \\"content\\": \\"width=device-width, initial-scale=1.0\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 10
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 11
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"meta\\",
                    \\"attributes\\": {
                      \\"http-equiv\\": \\"X-UA-Compatible\\",
                      \\"content\\": \\"ie=edge\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 12
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 13
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"title\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"Log record\\",
                        \\"id\\": 15
                      }
                    ],
                    \\"id\\": 14
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n    \\",
                    \\"id\\": 16
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"src\\": \\"http://localhost:3030/test/html/index.ts\\",
                      \\"type\\": \\"module\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 17
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n  \\",
                    \\"id\\": 18
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 3,
                \\"textContent\\": \\"\\\\n  \\",
                \\"id\\": 19
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n\\\\n\\",
                    \\"id\\": 21
                  }
                ],
                \\"id\\": 20
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"assert\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:2:15\\"
        ],
        \\"payload\\": [
          \\"true\\",
          \\"\\\\\\"assert\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"count\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:3:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"count\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"countReset\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:4:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"count\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"debug\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:5:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"debug\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"dir\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:6:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"dir\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"dirxml\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:7:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"dirxml\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"group\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:8:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"groupCollapsed\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:9:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"info\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:10:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"info\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"log\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:11:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"log\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"table\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:12:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"table\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"time\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:13:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"timeEnd\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:14:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"timeLog\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:15:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"trace\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:16:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"trace\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"warn\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:17:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"warn\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"clear\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:18:15\\"
        ],
        \\"payload\\": []
      }
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"log\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:19:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"TypeError: a message\\\\\\\\n    at __puppeteer_evaluation_script__:19:19\\\\\\\\nEnd of stack for Error object\\\\\\"\\"
        ]
      }
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"removes\\": [],
      \\"adds\\": [
        {
          \\"parentId\\": 20,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 2,
            \\"tagName\\": \\"iframe\\",
            \\"attributes\\": {},
            \\"childNodes\\": [],
            \\"id\\": 22
          }
        }
      ]
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"adds\\": [
        {
          \\"parentId\\": 22,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 0,
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"html\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"head\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [],
                    \\"rootId\\": 23,
                    \\"id\\": 25
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"body\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [],
                    \\"rootId\\": 23,
                    \\"id\\": 26
                  }
                ],
                \\"rootId\\": 23,
                \\"id\\": 24
              }
            ],
            \\"compatMode\\": \\"BackCompat\\",
            \\"id\\": 23
          }
        }
      ],
      \\"removes\\": [],
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"isAttachIframe\\": true
    }
  },
  {
    \\"type\\": 6,
    \\"data\\": {
      \\"plugin\\": \\"rrweb/console@1\\",
      \\"payload\\": {
        \\"level\\": \\"log\\",
        \\"trace\\": [
          \\"__puppeteer_evaluation_script__:2:15\\"
        ],
        \\"payload\\": [
          \\"\\\\\\"from iframe\\\\\\"\\"
        ]
      }
    }
  }
]"
`;
